src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,267 @@
+                    final int cardinality = resps[index].getCardinality();
+
+                    planWithStatistics = new PhysicalPlanWithStatistics( subplan, null, new ArrayList<>(), cardinality, 0 );
+                    index++;
+                }
+                else if ( pop instanceof PhysicalOpLocalToGlobal ){
+                    if (  subplan.getSubPlan(0).getRootOperator() instanceof PhysicalOpRequest
+                     || (subplan.getSubPlan(0).getRootOperator() instanceof PhysicalOpFilter && subplan.getSubPlan(0).getSubPlan(0).getRootOperator() instanceof PhysicalOpRequest)) {
+                        final int cardinality = resps[index].getCardinality();
+                        final VocabularyMapping vm = ((LogicalOpLocalToGlobal) ((PhysicalOpLocalToGlobal) pop).getLogicalOperator()).getVocabularyMapping();
+
+                        planWithStatistics = new PhysicalPlanWithStatistics(subplan, null, Arrays.asList(vm), cardinality, 0 );
+                        index++;
+                    }
+                }
+                else if (pop instanceof PhysicalOpBinaryUnion || pop instanceof PhysicalOpMultiwayUnion) {
+                    int aggregatedCardinality = 0;
+                    final List<VocabularyMapping> vms = new ArrayList<>();
+                    for ( int count = 0; count < subplan.numberOfSubPlans(); count++ ) {
+                        final int cardinality = resps[index].getCardinality();
+                        aggregatedCardinality += (cardinality < 0 ) ? Integer.MAX_VALUE : cardinality;
+                        if ( aggregatedCardinality < 0 ) aggregatedCardinality = Integer.MAX_VALUE;
+
+                        if ( subplan.getSubPlan(count).getRootOperator() instanceof PhysicalOpLocalToGlobal){
+                            final VocabularyMapping vm = ((LogicalOpLocalToGlobal) ((PhysicalOpLocalToGlobal) subplan.getSubPlan(count).getRootOperator()).getLogicalOperator()).getVocabularyMapping();
+                            vms.add(vm);
+                        }
+
+                        index++;
+                    }
+
+                    planWithStatistics = new PhysicalPlanWithStatistics(subplan, null, vms, aggregatedCardinality, 0 );
+                }
+                else
+                    throw new IllegalArgumentException("Unsupported type of subquery in source assignment (" + pop.getClass().getName() + ")");
+
+                subPlansWithStatistics.add( planWithStatistics );
+            }
+
+            return subPlansWithStatistics;
+        }
+
+        /**
+         * Construct a physical plan using symmetric hash join:
+         * If both currentPlan and nextPlan have PhysicalOpLocalToGlobal as root operator, and use the same vocabulary mapping,
+         * the physical plan can be simplified as:
+         * join( l2g^{vm}(p1), l2g^{vm}(p2) ) --> l2g^{vm}( join (p1, p2) )
```

Comment:
I don't think that this rewriting is correct in general. Here is a counterexample: Assume a vocab.mapping with

Local1  rdfs:subClassOf  Global
Local2  rdfs:subClassOf  Global

Assume two solution mappings:

m1 = { ?x -> Local1 }
m2 = { ?x -> Local2 }

where m1 comes from p1 and m2 comes from p2.

When applying the vocab.mapping directly to m1 and m2 (before doing the join), then the rewritten solution mappings are:

m1' = { ?x -> Global }
m2' = { ?x -> Global }

These are compatible and, thus, they join up with one another.

In contrast, when doing the join first, m1 and m2 are not compatible and, thus, they cannot be joined.

Reply:
Thank you very much for your explanation. Indeed! This rewriting is incorrect. 
I have removed it.